File: backend/app/schemas/review_cycle.py
"""Pydantic schemas for Phase 13 multi-round resubmittal and revision cycle.

These schemas cover review cycles, resubmittal packages and documents, applicant
responses and mappings, DXF metadata revision comparison, issue carry-forward,
response resolution, and next-cycle preparation. Revision comparison compares
extracted DXF metadata only. Nothing here verifies CAD, validates design,
certifies compliance, or makes a final engineering decision.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# Review cycle.


class ReviewCycleCreate(BaseModel):
    cycle_number: int | None = None
    cycle_name: str | None = None
    source_response_package_id: str | None = None
    source_workflow_board_id: str | None = None
    summary: str | None = None


class ReviewCycleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    review_cycle_id: str
    project_id: str
    cycle_number: int
    cycle_name: str
    status: str
    started_at: datetime
    completed_at: datetime | None = None
    source_response_package_id: str | None = None
    source_workflow_board_id: str | None = None
    summary: str
    requires_human_review: bool
    created_at: datetime
    updated_at: datetime


class ReviewCycleSummary(BaseModel):
    project_id: str
    cycle_count: int
    active_cycle_id: str | None = None
    active_cycle_number: int | None = None
    statuses: dict[str, int] = Field(default_factory=dict)
    note: str


# Resubmittal package and documents.


class ResubmittalPackageCreate(BaseModel):
    review_cycle_id: str | None = None
    package_name: str
    submitted_by: str = "applicant"
    summary: str | None = None


class ResubmittalPackageStatusUpdate(BaseModel):
    status: str
    reviewer_note: str | None = None


class ResubmittalDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    resubmittal_document_id: str
    project_id: str
    review_cycle_id: str
    resubmittal_package_id: str
    document_type: str
    source_type: str
    source_id: str | None = None
    file_name: str | None = None
    description: str
    status: str
    created_at: datetime


class ResubmittalPackageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    resubmittal_package_id: str
    project_id: str
    review_cycle_id: str
    package_name: str
    submitted_by: str
    submitted_at: datetime | None = None
    received_at: datetime
    status: str
    summary: str
    reviewer_note: str | None = None
    requires_human_review: bool
    created_at: datetime
    updated_at: datetime


class ResubmittalPackageDetail(ResubmittalPackageRead):
    documents: list[ResubmittalDocumentRead] = Field(default_factory=list)
    applicant_responses: list["ApplicantResponseRead"] = Field(
        default_factory=list
    )
    note: str = ""


# Applicant responses and mappings.


class ApplicantResponseCreate(BaseModel):
    response_text: str
    response_topic: str = "general"
    submitted_by: str = "applicant"
    target_response_item_id: str | None = None
    target_workflow_item_id: str | None = None


class ApplicantResponseRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    applicant_response_id: str
    project_id: str
    review_cycle_id: str
    resubmittal_package_id: str
    response_text: str
    response_topic: str
    submitted_by: str
    target_response_item_id: str | None = None
    target_workflow_item_id: str | None = None
    status: str
    reviewer_note: str | None = None
    requires_human_review: bool
    created_at: datetime
    updated_at: datetime


class ApplicantResponseMappingCreate(BaseModel):
    response_package_item_id: str | None = None
    workflow_item_id: str | None = None
    mapping_confidence: str = "medium"
    mapping_reason: str | None = None


class ApplicantResponseMappingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    mapping_id: str
    project_id: str
    review_cycle_id: str
    applicant_response_id: str
    response_package_item_id: str | None = None
    workflow_item_id: str | None = None
    response_resolution_record_id: str | None = None
    mapping_confidence: str
    mapping_reason: str
    requires_human_review: bool
    created_at: datetime


class ResponseMappingSummary(BaseModel):
    review_cycle_id: str
    project_id: str
    response_count: int
    mapped_count: int
    unmapped_count: int
    suggested_count: int
    confidence_counts: dict[str, int] = Field(default_factory=dict)
    note: str


# Revision comparison.


class RevisionComparisonCreate(BaseModel):
    previous_parse_run_id: str
    current_parse_run_id: str
    resubmittal_package_id: str | None = None


class RevisionChangeRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    change_record_id: str
    project_id: str
    review_cycle_id: str
    comparison_run_id: str
    change_type: str
    source_category: str
    previous_value: str | None = None
    current_value: str | None = None
    normalized_key: str
    layer_name: str | None = None
    reference_type: str | None = None
    severity: str
    linked_cad_review_finding_id: str | None = None
    linked_workflow_item_id: str | None = None
    reviewer_status: str
    reviewer_note: str | None = None
    requires_human_review: bool
    created_at: datetime


class RevisionComparisonRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    comparison_run_id: str
    project_id: str
    review_cycle_id: str
    resubmittal_package_id: str | None = None
    previous_parse_run_id: str
    current_parse_run_id: str
    status: str
    started_at: datetime
    completed_at: datetime | None = None
    compared_layer_count: int
    compared_text_count: int
    added_count: int
    removed_count: int
    changed_count: int
    unchanged_count: int
    warning_count: int
    summary: str
    limitations_note: str
    requires_human_review: bool


class RevisionComparisonSummary(BaseModel):
    comparison_run_id: str
    project_id: str
    review_cycle_id: str
    status: str
    added_count: int
    removed_count: int
    changed_count: int
    unchanged_count: int
    carried_forward_count: int
    changes_by_category: dict[str, int] = Field(default_factory=dict)
    changes_by_type: dict[str, int] = Field(default_factory=dict)
    limitations_note: str
    note: str


# Issue carry-forward.


class IssueCarryForwardCreate(BaseModel):
    source_workflow_item_id: str | None = None
    source_response_item_id: str | None = None
    source_cad_finding_id: str | None = None
    source_revision_change_id: str | None = None
    title: str | None = None
    reason: str
    reviewer_name: str = "reviewer"
    reviewer_note: str | None = None


class IssueCarryForwardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    carry_forward_id: str
    project_id: str
    review_cycle_id: str
    source_workflow_item_id: str | None = None
    source_response_item_id: str | None = None
    source_cad_finding_id: str | None = None
    source_revision_change_id: str | None = None
    title: str
    reason: str
    carried_forward_status: str
    target_workflow_item_id: str | None = None
    created_at: datetime
    reviewer_name: str
    reviewer_note: str | None = None
    requires_human_review: bool


class CarryForwardResult(BaseModel):
    review_cycle_id: str
    project_id: str
    created_count: int
    skipped_count: int
    carry_forward_ids: list[str] = Field(default_factory=list)
    note: str


class CarryForwardSummary(BaseModel):
    review_cycle_id: str
    project_id: str
    total: int
    statuses: dict[str, int] = Field(default_factory=dict)
    note: str


# Response resolution.


class ResponseResolutionCreate(BaseModel):
    response_package_item_id: str | None = None
    workflow_item_id: str | None = None
    applicant_response_id: str | None = None
    revision_change_record_id: str | None = None
    status: str = "still_open"
    reviewer_note: str | None = None
    reviewer_name: str = "reviewer"


class ResponseResolutionStatusUpdate(BaseModel):
    status: str
    reviewer_note: str | None = None
    reviewer_name: str = "reviewer"


class ResponseResolutionRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    resolution_record_id: str
    project_id: str
    review_cycle_id: str
    response_package_item_id: str | None = None
    workflow_item_id: str | None = None
    applicant_response_id: str | None = None
    revision_change_record_id: str | None = None
    status: str
    reviewer_note: str | None = None
    reviewer_name: str
    created_at: datetime
    updated_at: datetime
    requires_human_review: bool


class ResolutionSummary(BaseModel):
    review_cycle_id: str
    project_id: str
    total: int
    statuses: dict[str, int] = Field(default_factory=dict)
    note: str


# Next-cycle preparation.


class NextCyclePreparationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    next_cycle_preparation_id: str
    project_id: str
    review_cycle_id: str
    status: str
    summary: str
    carried_forward_count: int
    needs_more_information_count: int
    reviewer_checked_count: int
    next_response_package_ready: bool
    created_at: datetime
    updated_at: datetime
    requires_human_review: bool


# Dashboard.


class ReviewCycleDashboard(BaseModel):
    project_id: str
    cycle_count: int
    active_cycle_id: str | None = None
    active_cycle_number: int | None = None
    review_cycles: list[ReviewCycleRead] = Field(default_factory=list)
    resubmittal_count: int
    resubmittal_statuses: dict[str, int] = Field(default_factory=dict)
    applicant_response_count: int
    unmapped_response_count: int
    comparison_run_count: int
    revision_change_count: int
    carry_forward_count: int
    resolution_count: int
    resolution_statuses: dict[str, int] = Field(default_factory=dict)
    open_item_count: int
    next_cycle_ready: bool
    limitations_note: str


ResubmittalPackageDetail.model_rebuild()

File: lib/api/reviewCycle.ts
import {
  API_BASE_URL,
  PROJECT_ID,
  apiFetch,
  apiGetMapped,
  authHeaders,
  requireString,
  type ApiResult,
} from "./client";

// Phase 13: multi-round resubmittal, revision comparison, and applicant response
// cycle. Data is backend-canonical. The frontend does not simulate review cycle
// data. Read calls return a typed ApiResult that preserves the failure category,
// and mutating calls return a clear backend-required result.

export type ReviewCycle = {
  reviewCycleId: string;
  projectId: string;
  cycleNumber: number;
  cycleName: string;
  status: string;
  startedAt: string;
  completedAt: string | null;
  sourceResponsePackageId: string | null;
  sourceWorkflowBoardId: string | null;
  summary: string;
  requiresHumanReview: boolean;
  createdAt: string;
  updatedAt: string;
};

export type ResubmittalDocument = {
  resubmittalDocumentId: string;
  projectId: string;
  reviewCycleId: string;
  resubmittalPackageId: string;
  documentType: string;
  sourceType: string;
  sourceId: string | null;
  fileName: string | null;
  description: string;
  status: string;
  createdAt: string;
};

export type ApplicantResponse = {
  applicantResponseId: string;
  projectId: string;
  reviewCycleId: string;
  resubmittalPackageId: string;
  responseText: string;
  responseTopic: string;
  submittedBy: string;
  targetResponseItemId: string | null;
  targetWorkflowItemId: string | null;
  status: string;
  reviewerNote: string | null;
  requiresHumanReview: boolean;
  createdAt: string;
  updatedAt: string;
};

export type ResubmittalPackage = {
  resubmittalPackageId: string;
  projectId: string;
  reviewCycleId: string;
  packageName: string;
  submittedBy: string;
  submittedAt: string | null;
  receivedAt: string;
  status: string;
  summary: string;
  reviewerNote: string | null;
  requiresHumanReview: boolean;
  createdAt: string;
  updatedAt: string;
  documents?: ResubmittalDocument[];
  applicantResponses?: ApplicantResponse[];
};

export type ApplicantResponseMapping = {
  mappingId: string;
  projectId: string;
  reviewCycleId: string;
  applicantResponseId: string;
  responsePackageItemId: string | null;
  workflowItemId: string | null;
  responseResolutionRecordId: string | null;
  mappingConfidence: string;
  mappingReason: string;
  requiresHumanReview: boolean;
  createdAt: string;
};

export type RevisionComparisonRun = {
  comparisonRunId: string;
  projectId: string;
  reviewCycleId: string;
  resubmittalPackageId: string | null;
  previousParseRunId: string;
  currentParseRunId: string;
  status: string;
  startedAt: string;
  completedAt: string | null;
  comparedLayerCount: number;
  comparedTextCount: number;
  addedCount: number;
  removedCount: number;
  changedCount: number;
  unchangedCount: number;
  warningCount: number;
  summary: string;
  limitationsNote: string;
  requiresHumanReview: boolean;
};

export type RevisionChangeRecord = {
  changeRecordId: string;
  projectId: string;
  reviewCycleId: string;
  comparisonRunId: string;
  changeType: string;
  sourceCategory: string;
  previousValue: string | null;
  currentValue: string | null;
  normalizedKey: string;
  layerName: string | null;
  referenceType: string | null;
  severity: string;
  linkedCadReviewFindingId: string | null;
  linkedWorkflowItemId: string | null;
  reviewerStatus: string;
  reviewerNote: string | null;
  requiresHumanReview: boolean;
  createdAt: string;
};

export type IssueCarryForward = {
  carryForwardId: string;
  projectId: string;
  reviewCycleId: string;
  sourceWorkflowItemId: string | null;
  sourceResponseItemId: string | null;
  sourceCadFindingId: string | null;
  sourceRevisionChangeId: string | null;
  title: string;
  reason: string;
  carriedForwardStatus: string;
  targetWorkflowItemId: string | null;
  createdAt: string;
  reviewerName: string;
  reviewerNote: string | null;
  requiresHumanReview: boolean;
};

export type ResponseResolutionRecord = {
  resolutionRecordId: string;
  projectId: string;
  reviewCycleId: string;
  responsePackageItemId: string | null;
  workflowItemId: string | null;
  applicantResponseId: string | null;
  revisionChangeRecordId: string | null;
  status: string;
  reviewerNote: string | null;
  reviewerName: string;
  createdAt: string;
  updatedAt: string;
  requiresHumanReview: boolean;
};

export type NextCyclePreparation = {
  nextCyclePreparationId: string;
  projectId: string;
  reviewCycleId: string;
  status: string;
  summary: string;
  carriedForwardCount: number;
  needsMoreInformationCount: number;
  reviewerCheckedCount: number;
  nextResponsePackageReady: boolean;
  createdAt: string;
  updatedAt: string;
  requiresHumanReview: boolean;
};

export type ReviewCycleDashboard = {
  projectId: string;
  cycleCount: number;
  activeCycleId: string | null;
  activeCycleNumber: number | null;
  reviewCycles: ReviewCycle[];
  resubmittalCount: number;
  resubmittalStatuses: Record<string, number>;
  applicantResponseCount: number;
  unmappedResponseCount: number;
  comparisonRunCount: number;
  revisionChangeCount: number;
  carryForwardCount: number;
  resolutionCount: number;
  resolutionStatuses: Record<string, number>;
  openItemCount: number;
  nextCycleReady: boolean;
  limitationsNote: string;
};

export type ResponseMappingSummary = {
  reviewCycleId: string;
  projectId: string;
  responseCount: number;
  mappedCount: number;
  unmappedCount: number;
  suggestedCount: number;
  confidenceCounts: Record<string, number>;
  note: string;
};

export type RevisionComparisonSummary = {
  comparisonRunId: string;
  projectId: string;
  reviewCycleId: string;
  status: string;
  addedCount: number;
  removedCount: number;
  changedCount: number;
  unchangedCount: number;
  carriedForwardCount: number;
  changesByCategory: Record<string, number>;
  changesByType: Record<string, number>;
  limitationsNote: string;
  note: string;
};

export type CarryForwardSummary = {
  reviewCycleId: string;
  projectId: string;
  total: number;
  statuses: Record<string, number>;
  note: string;
};

export type ResolutionSummary = {
  reviewCycleId: string;
  projectId: string;
  total: number;
  statuses: Record<string, number>;
  note: string;
};

type Json = Record<string, unknown>;

function camel<T>(obj: Json | null | undefined): T {
  const out: Json = {};
  if (!obj) return out as T;
  for (const [key, value] of Object.entries(obj)) {
    const ck = key.replace(/_([a-z])/g, (_m, c: string) => c.toUpperCase());
    out[ck] = value;
  }
  return out as T;
}

function mapReviewCycle(d: Json): ReviewCycle {
  return camel<ReviewCycle>(d);
}

// Strict read-path mapper. The requireString assertions make a structurally
// invalid backend payload surface as an explicit invalid_response failure
// through apiGetMapped instead of propagating undefined fields into the UI.
function mapReviewCycleRead(d: Json): ReviewCycle {
  requireString(d.review_cycle_id, "review_cycle_id");
  requireString(d.project_id, "project_id");
  return mapReviewCycle(d);
}

async function postJson<T>(
  path: string,
  body: unknown,
  errorPrefix: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: T; error?: string }> {
  try {
    const res = await fetch(`${API_BASE_URL}${path}`, {
      method: "POST",
      headers: authHeaders(
        body === undefined ? undefined : { "Content-Type": "application/json" },
      ),
      body: body === undefined ? undefined : JSON.stringify(body),
      cache: "no-store",
    });
    if (!res.ok) {
      let detail = `Request failed (${res.status}).`;
      try {
        const errBody = (await res.json()) as { detail?: string };
        if (errBody.detail) detail = errBody.detail;
      } catch {
        // keep generic
      }
      return { ok: false, backendReachable: true, error: detail };
    }
    return { ok: true, backendReachable: true, data: (await res.json()) as T };
  } catch {
    return {
      ok: false,
      backendReachable: false,
      error: `${errorPrefix} The backend is not reachable. Review cycle data is not simulated in the browser.`,
    };
  }
}

async function patchJson<T>(
  path: string,
  body: unknown,
): Promise<{ ok: boolean; backendReachable: boolean; data?: T; error?: string }> {
  try {
    const res = await fetch(`${API_BASE_URL}${path}`, {
      method: "PATCH",
      headers: authHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify(body),
      cache: "no-store",
    });
    if (!res.ok) {
      let detail = `Request failed (${res.status}).`;
      try {
        const errBody = (await res.json()) as { detail?: string };
        if (errBody.detail) detail = errBody.detail;
      } catch {
        // keep generic
      }
      return { ok: false, backendReachable: true, error: detail };
    }
    return { ok: true, backendReachable: true, data: (await res.json()) as T };
  } catch {
    return { ok: false, backendReachable: false, error: "The backend is not reachable." };
  }
}

// Review cycle.

export async function getReviewCycles(): Promise<ApiResult<ReviewCycle[]>> {
  return apiGetMapped<Json[], ReviewCycle[]>(
    `/api/v1/projects/${PROJECT_ID}/review-cycles`,
    (data) => data.map(mapReviewCycleRead),
  );
}

export async function getReviewCycle(
  reviewCycleId: string,
): Promise<ApiResult<ReviewCycle>> {
  return apiGetMapped<Json, ReviewCycle>(
    `/api/v1/review-cycles/${reviewCycleId}`,
    mapReviewCycleRead,
  );
}

export async function getReviewCycleDashboard(): Promise<ApiResult<ReviewCycleDashboard>> {
  return apiGetMapped<Json, ReviewCycleDashboard>(
    `/api/v1/projects/${PROJECT_ID}/review-cycle-dashboard`,
    (data) => {
      const mapped = camel<ReviewCycleDashboard>(data);
      mapped.reviewCycles = ((data.review_cycles as Json[]) ?? []).map(
        mapReviewCycleRead,
      );
      return mapped;
    },
  );
}

export async function getReviewCycleSummary(): Promise<ApiResult<Json>> {
  return apiFetch<Json>(
    `/api/v1/projects/${PROJECT_ID}/review-cycle-summary`,
  );
}

export async function createReviewCycle(
  cycleName?: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: ReviewCycle; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/projects/${PROJECT_ID}/review-cycles`,
    { cycle_name: cycleName ?? null },
    "Could not create the review cycle.",
  );
  return { ...result, data: result.data ? mapReviewCycle(result.data) : undefined };
}

// Resubmittal.

export async function getResubmittalPackages(): Promise<ApiResult<ResubmittalPackage[]>> {
  return apiGetMapped<Json[], ResubmittalPackage[]>(
    `/api/v1/projects/${PROJECT_ID}/resubmittals`,
    (data) => data.map((d) => camel<ResubmittalPackage>(d)),
  );
}

export async function getResubmittalPackage(
  resubmittalPackageId: string,
): Promise<ApiResult<ResubmittalPackage>> {
  return apiGetMapped<Json, ResubmittalPackage>(
    `/api/v1/resubmittals/${resubmittalPackageId}`,
    (data) => {
      const mapped = camel<ResubmittalPackage>(data);
      mapped.documents = ((data.documents as Json[]) ?? []).map((d) =>
        camel<ResubmittalDocument>(d),
      );
      mapped.applicantResponses = (
        (data.applicant_responses as Json[]) ?? []
      ).map((d) => camel<ApplicantResponse>(d));
      return mapped;
    },
  );
}

export async function createResubmittalPackage(
  reviewCycleId: string,
  packageName: string,
  submittedBy = "applicant",
): Promise<{ ok: boolean; backendReachable: boolean; data?: ResubmittalPackage; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/projects/${PROJECT_ID}/resubmittals`,
    {
      review_cycle_id: reviewCycleId,
      package_name: packageName,
      submitted_by: submittedBy,
    },
    "Could not create the resubmittal package.",
  );
  return {
    ...result,
    data: result.data ? camel<ResubmittalPackage>(result.data) : undefined,
  };
}

export async function updateResubmittalPackageStatus(
  resubmittalPackageId: string,
  status: string,
  reviewerNote?: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: ResubmittalPackage; error?: string }> {
  const result = await patchJson<Json>(
    `/api/v1/resubmittals/${resubmittalPackageId}/status`,
    { status, reviewer_note: reviewerNote ?? null },
  );
  return {
    ...result,
    data: result.data ? camel<ResubmittalPackage>(result.data) : undefined,
  };
}

export async function linkCadFileToResubmittal(
  resubmittalPackageId: string,
  cadFileId: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: ResubmittalPackage; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/resubmittals/${resubmittalPackageId}/cad-files/${cadFileId}`,
    undefined,
    "Could not link the CAD file.",
  );
  return {
    ...result,
    data: result.data ? camel<ResubmittalPackage>(result.data) : undefined,
  };
}

export async function createApplicantResponse(
  resubmittalPackageId: string,
  responseText: string,
  responseTopic = "general",
  submittedBy = "applicant",
): Promise<{ ok: boolean; backendReachable: boolean; data?: ApplicantResponse; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/resubmittals/${resubmittalPackageId}/applicant-responses`,
    {
      response_text: responseText,
      response_topic: responseTopic,
      submitted_by: submittedBy,
    },
    "Could not add the applicant response.",
  );
  return {
    ...result,
    data: result.data ? camel<ApplicantResponse>(result.data) : undefined,
  };
}

// Applicant responses and mappings.

export async function getApplicantResponses(): Promise<ApiResult<ApplicantResponse[]>> {
  return apiGetMapped<Json[], ApplicantResponse[]>(
    `/api/v1/projects/${PROJECT_ID}/applicant-responses`,
    (data) => data.map((d) => camel<ApplicantResponse>(d)),
  );
}

export async function createApplicantResponseMapping(
  applicantResponseId: string,
  options: {
    responsePackageItemId?: string;
    workflowItemId?: string;
    mappingConfidence?: string;
    mappingReason?: string;
  },
): Promise<{ ok: boolean; backendReachable: boolean; data?: ApplicantResponseMapping; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/applicant-responses/${applicantResponseId}/mappings`,
    {
      response_package_item_id: options.responsePackageItemId ?? null,
      workflow_item_id: options.workflowItemId ?? null,
      mapping_confidence: options.mappingConfidence ?? "medium",
      mapping_reason: options.mappingReason ?? null,
    },
    "Could not create the mapping.",
  );
  return {
    ...result,
    data: result.data ? camel<ApplicantResponseMapping>(result.data) : undefined,
  };
}

export async function suggestResponseMappings(
  reviewCycleId: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: ApplicantResponseMapping[]; error?: string }> {
  const result = await postJson<Json[]>(
    `/api/v1/review-cycles/${reviewCycleId}/suggest-response-mappings`,
    undefined,
    "Could not suggest mappings.",
  );
  return {
    ...result,
    data: result.data
      ? result.data.map((d) => camel<ApplicantResponseMapping>(d))
      : undefined,
  };
}

export async function getResponseMappingSummary(
  reviewCycleId: string,
): Promise<ApiResult<ResponseMappingSummary>> {
  return apiGetMapped<Json, ResponseMappingSummary>(
    `/api/v1/review-cycles/${reviewCycleId}/response-mapping-summary`,
    (data) => camel<ResponseMappingSummary>(data),
  );
}

// Revision comparison.

export async function runRevisionComparison(
  reviewCycleId: string,
  previousParseRunId: string,
  currentParseRunId: string,
  resubmittalPackageId?: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: RevisionComparisonRun; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/review-cycles/${reviewCycleId}/revision-comparisons`,
    {
      previous_parse_run_id: previousParseRunId,
      current_parse_run_id: currentParseRunId,
      resubmittal_package_id: resubmittalPackageId ?? null,
    },
    "Could not run the revision comparison.",
  );
  return {
    ...result,
    data: result.data ? camel<RevisionComparisonRun>(result.data) : undefined,
  };
}

export async function getRevisionComparisons(): Promise<ApiResult<RevisionComparisonRun[]>> {
  return apiGetMapped<Json[], RevisionComparisonRun[]>(
    `/api/v1/projects/${PROJECT_ID}/revision-comparisons`,
    (data) => data.map((d) => camel<RevisionComparisonRun>(d)),
  );
}

export async function getRevisionComparison(
  comparisonRunId: string,
): Promise<ApiResult<RevisionComparisonRun>> {
  return apiGetMapped<Json, RevisionComparisonRun>(
    `/api/v1/revision-comparisons/${comparisonRunId}`,
    (data) => camel<RevisionComparisonRun>(data),
  );
}

export async function getRevisionChanges(
  comparisonRunId: string,
): Promise<ApiResult<RevisionChangeRecord[]>> {
  return apiGetMapped<Json[], RevisionChangeRecord[]>(
    `/api/v1/revision-comparisons/${comparisonRunId}/changes`,
    (data) => data.map((d) => camel<RevisionChangeRecord>(d)),
  );
}

export async function getRevisionComparisonSummary(
  comparisonRunId: string,
): Promise<ApiResult<RevisionComparisonSummary>> {
  return apiGetMapped<Json, RevisionComparisonSummary>(
    `/api/v1/revision-comparisons/${comparisonRunId}/summary`,
    (data) => camel<RevisionComparisonSummary>(data),
  );
}

// Carry-forward.

export async function carryForwardUnresolvedItems(
  reviewCycleId: string,
): Promise<{ ok: boolean; backendReachable: boolean; createdCount?: number; skippedCount?: number; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/review-cycles/${reviewCycleId}/carry-forward`,
    undefined,
    "Could not carry items forward.",
  );
  return {
    ok: result.ok,
    backendReachable: result.backendReachable,
    error: result.error,
    createdCount: result.data?.created_count as number | undefined,
    skippedCount: result.data?.skipped_count as number | undefined,
  };
}

export async function getCarryForwards(): Promise<ApiResult<IssueCarryForward[]>> {
  return apiGetMapped<Json[], IssueCarryForward[]>(
    `/api/v1/projects/${PROJECT_ID}/carry-forwards`,
    (data) => data.map((d) => camel<IssueCarryForward>(d)),
  );
}

export async function getCarryForwardSummary(
  reviewCycleId: string,
): Promise<ApiResult<CarryForwardSummary>> {
  return apiGetMapped<Json, CarryForwardSummary>(
    `/api/v1/review-cycles/${reviewCycleId}/carry-forward-summary`,
    (data) => camel<CarryForwardSummary>(data),
  );
}

// Resolution.

export async function createResponseResolutionRecord(
  reviewCycleId: string,
  status: string,
  options: {
    responsePackageItemId?: string;
    workflowItemId?: string;
    applicantResponseId?: string;
    revisionChangeRecordId?: string;
    reviewerNote?: string;
    reviewerName?: string;
  } = {},
): Promise<{ ok: boolean; backendReachable: boolean; data?: ResponseResolutionRecord; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/review-cycles/${reviewCycleId}/resolution-records`,
    {
      response_package_item_id: options.responsePackageItemId ?? null,
      workflow_item_id: options.workflowItemId ?? null,
      applicant_response_id: options.applicantResponseId ?? null,
      revision_change_record_id: options.revisionChangeRecordId ?? null,
      status,
      reviewer_note: options.reviewerNote ?? null,
      reviewer_name: options.reviewerName ?? "reviewer",
    },
    "Could not create the resolution record.",
  );
  return {
    ...result,
    data: result.data ? camel<ResponseResolutionRecord>(result.data) : undefined,
  };
}

export async function updateResponseResolutionStatus(
  resolutionRecordId: string,
  status: string,
  reviewerName = "reviewer",
  reviewerNote?: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: ResponseResolutionRecord; error?: string }> {
  const result = await patchJson<Json>(
    `/api/v1/resolution-records/${resolutionRecordId}/status`,
    { status, reviewer_name: reviewerName, reviewer_note: reviewerNote ?? null },
  );
  return {
    ...result,
    data: result.data ? camel<ResponseResolutionRecord>(result.data) : undefined,
  };
}

export async function getResponseResolutionRecords(): Promise<ApiResult<ResponseResolutionRecord[]>> {
  return apiGetMapped<Json[], ResponseResolutionRecord[]>(
    `/api/v1/projects/${PROJECT_ID}/resolution-records`,
    (data) => data.map((d) => camel<ResponseResolutionRecord>(d)),
  );
}

export async function getResolutionSummary(
  reviewCycleId: string,
): Promise<ApiResult<ResolutionSummary>> {
  return apiGetMapped<Json, ResolutionSummary>(
    `/api/v1/review-cycles/${reviewCycleId}/resolution-summary`,
    (data) => camel<ResolutionSummary>(data),
  );
}

// Next cycle.

export async function prepareNextCycle(
  reviewCycleId: string,
): Promise<{ ok: boolean; backendReachable: boolean; data?: NextCyclePreparation; error?: string }> {
  const result = await postJson<Json>(
    `/api/v1/review-cycles/${reviewCycleId}/prepare-next-cycle`,
    undefined,
    "Could not prepare the next cycle.",
  );
  return {
    ...result,
    data: result.data ? camel<NextCyclePreparation>(result.data) : undefined,
  };
}

export async function getNextCyclePreparation(
  reviewCycleId: string,
): Promise<ApiResult<NextCyclePreparation>> {
  return apiGetMapped<Json, NextCyclePreparation>(
    `/api/v1/review-cycles/${reviewCycleId}/next-cycle-preparation`,
    (data) => camel<NextCyclePreparation>(data),
  );
}
